enhancements/validate-catalog-for-bad-bundles.md

```diff
@@ -0,0 +1,217 @@
+$ cue import pdb.yaml
+
+apiVersion: "policy/v1beta1"
+kind:       "PodDisruptionBudget"
+metadata: name: "zk-pdb"
+spec: {
+	minAvailable: >0 & <=7
+	selector: matchLabels: app: "zookeeper"
+}
+```
+- For example, a PDB cannot have it's `minAvailable` set to 0. We could enforce this restriction as shown in the above yaml
+> **_NOTE:_**  The generation of .cue file and persisting them in a directory is done once when a new k8 object is included as part of the bundle (This is part of the initial setup)
+
+#### Validate YAML against the generated `.cue` 
+- The `vet` command of the cue command line tool can validate YAML files using a CUE schema
+```yaml
+$ cue vet pdb.yaml pdb.cue
+
+spec.minAvailable: invalid value 8 (out of bound <=7):
+    ./pdb.cue:5:21
+    ./pdb.yaml:6:18
+```
+- These .cue templates could be stored in a directory to be consumed by `opm` sub command (detailed in part 2)
+-  The basic idea is to unpack the catalog image, and for each bundle image, unpack it's content into a temp dir.
+-  We then identify the gvk type(k8 object, say PDB) from the manifest and run `cue vet <pdb.cue> <pdb.yaml>` which returns a `bool` with an error message.
+
+
+#### Pros
+- Reduce boilerplate code 
+- Not hinging on a third-party operator like OPA Gatekeeper.
+- Validation is extendable to customized .cue files
+- As a future scope, CUE can be wrapped around using [Tilt](https://tilt.dev/) to be running against a k8 cluster as we make changes to the cue configuration files. [More details here](https://garethr.dev/2019/04/automating-the-cue-workflow-with-tilt/)
+- Cue also allows you to define custom cmd flags under CUE cli [(only mentioned in source code)](https://github.com/cuelang/cue/blob/2b0e7cd9f63a190e762d7c802b98528ff80dcb7c/cmd/cue/cmd/cmd.go#L31) but there is no documentation on this
+
+#### Cons
+- The set of rules are defined as part of the pipeline (pre-flight checks) and the cluster-admin has no customization option
+- Not a lot to offer in terms of flexibility as compared to [OpenPolicyAgent (OPA)](https://kubernetes.io/blog/2019/08/06/opa-gatekeeper-policy-and-governance-for-kubernetes/). With OPA, you can enforce custom policies for create, update and delete of Kubernetes objects without recompiling or reconfiguring the Kubernetes API server or even Kubernetes Admission Controllers. 
+- Require users supplying CUE validation template to be familiar with a new type of language.
+- Documentation is vague and doesn't have concrete implementation examples
+- The error messages are very basic and not customizable(AFAIK)
+
+
+## Part 2: Tool to list "bad" bundles in a catalog
+The basic approach will be to:
+- Extend [`opm`](https://github.com/operator-framework/operator-registry/blob/master/docs/design/opm-tooling.md) tooling to add two new commands:
+    - `opm index extract <catalog-image>` - pull and unpack catalog image
+    - `opm index validate <registry file location>` - validate all the bundle images in the registry file
```

Comment:
Could we also allow users to specify a `catalog-image` here?

Reply:
Do you mean as part of `opm index validate <registry file location>` ? I have updated the design doc to reuse modified version of `opm index export` and then call the newly defined `opm index validate <dowloaded-location>